src/v/cluster/bootstrap_backend.cc

```diff
@@ -186,4 +190,37 @@ bootstrap_backend::apply(bootstrap_cluster_cmd_data cmd) {
     co_return errc::success;
 }
 
+ss::future<> bootstrap_backend::fill_snapshot(controller_snapshot& snap) const {
+    snap.bootstrap.cluster_uuid = _cluster_uuid_applied;
+    co_return;
+}
+
+ss::future<> bootstrap_backend::apply_snapshot(
+  const raft::snapshot_metadata&, const controller_snapshot& snap) {
+    if (!snap.bootstrap.cluster_uuid) {
+        co_return;
+    }
+
+    if (
+      _cluster_uuid_applied
+      && _storage.local().get_cluster_uuid() == snap.bootstrap.cluster_uuid) {
+        co_return;
+    }
+
+    // we don't fill other cmd fields because members_manager and
+    // credentials_store will bootstrap themselves from their respective base
+    // tables.
+    bootstrap_cluster_cmd_data cmd{
+      .uuid = *snap.bootstrap.cluster_uuid,
+    };
+    auto ec = co_await apply(std::move(cmd));
```

Comment:
I think I would rather take a different approach here. For some stms applying a snapshot is just a case of applying a command with delta. For those stms that impement one-shot state change as a command that may not be the case (not sure if there are others like that besides `bootstap_backend`). So instead of trying to squeeze the snapshot into a command that doesn't fit it really well, I think it's reasonable in this case to apply the snapshot data directly from `apply_snapshot` even if that causes duplication of a few lines.

As a side effect of this change, it will be clear then how snapshot application deals with the bootstrap idempotency. As of now, snapshot won't be applied if `cluster_uuid` is already in place. Apparently idempotency error is a logic error in case of snapshot application, and should be treated differently (`runtime_error` or log(error) at least).